### PR TITLE
Fix and enable external tests

### PR DESCRIPTION
**Pull request by cameel**
Related to #10397 (though my PR currently enables them in PR checks rather than in the nightly run).

This PR enables actual test runs (in addition to the current compilation runs) for external projects from `solidity-external-tests`.

An integral part of this PR is the update of `solidity-external-tests` repos. Here are the links in case you want to take a look at my changes. In all cases I pulled the most recent changes from the main upstream branch, created a `070` variant (or `070_new` when `070` already existed), cherry-picked our older changes where possible and then fixed the remaining problems.
- [Gnosis v1: `development_070_new`](https://github.com/solidity-external-tests/safe-contracts/tree/development_070_new) (depends on [mock-contract `master_070_new`](https://github.com/solidity-external-tests/mock-contract/tree/master_070_new))
- [Gnosis v2: `v2_070_new`](https://github.com/solidity-external-tests/safe-contracts/tree/v2_070) (depends on [mock-contract `master_070_new`](https://github.com/solidity-external-tests/mock-contract/tree/master_070_new) and the new [util-contracts](https://github.com/solidity-external-tests/util-contracts) that I have forked into `solidity-external-contracts`)
- [OpenZeppelin: `master_070`](https://github.com/solidity-external-tests/openzeppelin-contracts/tree/master_070)
- [Colony: `develop_070_new`](https://github.com/solidity-external-tests/colonyNetwork/tree/develop_070_new) (depends on [dappsys `master_070`](https://github.com/solidity-external-tests/dappsys-monolithic/tree/master_070))

Status:
- [x] Colony
- [x] Gnosis v1 (compilation only; tests disabled due to `ProviderError`)
- [x] Gnosis v2
- [x] OpenZeppelin
- [x] ENS

**Note: Once this PR gets merged into `breaking` (as a part of the usual merge from `develop`) it needs to be immediately followed by merging #10465 into `breaking` too (and also #10465 will need a rebase before that). Otherwise tests on breaking won't succeed because they'll still be using the `070` branches from `solidity-external-tests`.**

### CODE REVIEW

File: test/externalTests/gnosis.sh
Comment (axic) [inline test/externalTests/gnosis.sh]: Don't you have write access to `solidity-external-tests`?

Comment (cameel) [reply test/externalTests/gnosis.sh]: I don't.

Comment (axic) [reply test/externalTests/gnosis.sh]: It seems @ekpyron is the owner on it -- can you add @cameel please?

Comment (cameel) [reply test/externalTests/gnosis.sh]: Thanks. Though it's not urgent. This link is just a temporary change that will be removed from the final PR. I was going to submit my fixes to `solidity-external-tests` in PRs like e.g. https://github.com/solidity-external-tests/colonyNetwork/pull/8.

**Comment by cameel** [comment]
I'm done for today. Here's the current status.

### Summary
I managed to bring both Gnosis and Colony to a state where I could get actual test failures (rather than crashes in migrations or missing/conflicting JS dependencies).

A lot of problems I had to deal with have already been resolved in the upstream repos. I think it's going to be easier to just port our changes in `solidity-external-tests`to the most recent versions of each than try to continue with the old unsupported code. Colony has recently updated to Solidity 0.7.3 so that would not even require tweaking the contracts (though @hrkrshnn tried to switch to that (https://github.com/ethereum/solidity/commit/e5a232410f3cd30783e43a1b7870d6cc66d71034) and it did not work out of the box due to issues with npm dependencies). Gnosis is on Solidity 0.5.x so that will be harder but probably still easier than hacking it to work with newer Truffle.

### Colony
- There was a bug in how we updated the `fallback` in `EtherRouter` proxy for 0.6.0 (https://github.com/solidity-external-tests/colonyNetwork/pull/8).
- Now migrations are working but 136 tests fail.
    - A lot of them due to missing `ganache-accounts.json`. It does get generated when I run tests manually on my machine so I need to investigate that further.
    - There are other failures too.

### Gnosis
- It too had proxies with incorrectly updated fallback.
- `eth-lightwallet` dependency needed to be updated to 4.0.0 due to an unresolved dependency issue in 3.0.1 (https://github.com/ConsenSys/eth-lightwallet/issues/214). Gnosis has already updated to that version (https://github.com/gnosis/safe-contracts/pull/195).
- There are problems with both newer and older Truffle:
    - The original dependency is on Truffle 4.1.14.
        - This version does not work with the current `solc-js` (`AssertionError [ERR_ASSERTION]: Error parsing /tmp/d/safe-contracts/contracts/GnosisSafe.sol: Invalid callback object specified.`)
    - Our [test/externalTests/gnosis.sh](https://github.com/ethereum/solidity/blob/develop/test/externalTests/gnosis.sh) overwrites the dependency with Truffle 5.0.42, probably for that reason.
        - Truffle provides the `web3` library as a global object and Gnosis expects an older version of it which has `web3.toWei()` rather than `web3.utils.toWei()` so migrations fail on Truffle 5.x. This was also fixed by https://github.com/gnosis/safe-contracts/pull/195.
        - Minor tweaks are needed in `4_deploy_master_copies.js` migration because newer Truffle is stricter about types and `0` does not work as `address` or `bytes`. Again, fixed by https://github.com/gnosis/safe-contracts/pull/195.
- I managed to run tests with Truffle 4 on bytecode compiled by Truffle 5. Surprisingly only 3 tests are failing.
    - Two of them due to `web3.eth.call()` returning empty data from a call to `GnosisSafe.requiredTxGas()`. I think this might be related to the fallback but I could not easily investigate without recompiling.

File: test/externalTests/gnosis.sh
Comment (cameel) [reply test/externalTests/gnosis.sh]: Actually these permissions would be handy now because I wanted to pull latest changes from colony into the test repo.

Maybe @chriseth is an owner there too?

**Comment by cameel** [comment]
I managed to get Gnosis working on their `v2` branch. The bad news is that the tests on this branch are completely gutted. There's just one actually. Running v1 would be better if we want the tests to be worth anything.

File: test/externalTests/gnosis.sh
Comment (ekpyron) [reply test/externalTests/gnosis.sh]: I just added you there @cameel - I also changed @axic and @chriseth to owners, not sure why that wasn't the case so far. Arguably, we could also just make everyone in the team owner there, not sure...

Comment (cameel) [reply test/externalTests/gnosis.sh]: Thanks!

Not sure about owners - limiting permissions is often a good idea and for example I currently don't need that much - but giving everyone on the team write access sounds good.

**Review by chriseth — DISMISSED** [review]

**Comment by cameel** [comment]
I'm finished with making changes related to compiler version incompatibilities and this PR is *almost* ready to merge. *Almost* because Gnosis and OpenZeppelin tests seem to have some non-determinism and sometimes fail and sometimes pass. I spent a lot of time tweaking them but haven't been able to solve that yet. In this state merging them is risky - unless we're fine with them randomly failing in PRs.

I don't think these failures are related to changes in the compiler at all. In Gnosis it's either [`ProviderError` (Truffle loses connection to Ganache during a random test)](https://app.circleci.com/pipelines/github/ethereum/solidity/10994/workflows/d252d7d9-f24e-4cbb-8aee-b47480532aeb/jobs/530251) or [one of the tests failing due to invalid owner key](https://app.circleci.com/pipelines/github/ethereum/solidity/10979/workflows/f594ceef-8812-48a9-ac65-6ce0915451d9/jobs/529704). In case of OpenZeppelin it's a [timeout](https://app.circleci.com/pipelines/github/ethereum/solidity/10994/workflows/d252d7d9-f24e-4cbb-8aee-b47480532aeb/jobs/530252). External tests actually do 3 separate runs (each time with different optimization settings) and tests sometimes work flawlessly in some of the runs just to fail in another. Unfortunately even the trick with running them twice does not consistently help.

Another thing is that if we want to be able to freely merge `develop` into `breaking` and have tests on `breaking` pass, we should have 0.8.0 versions of all the tests ready before we merge this PR. Here's my draft PR for that: #10465. The only project that currently does not pass tests on `breaking` is OpenZeppelin. It needs fixes for safe math - tests do not pass on `upgrade-0.8.0` branch. Fixes are probably simple but I discovered it towards the end and I don't have energy for fixing them today. I'll try tomorrow. I focused on Gnosis and Colony and didn't expect to be backstabbed from that direction :)

So here's the current plan regarding this and related PRs:
- Merge #10383 "Force ENS external test to use latest Truffle" (fix for broken ENS on `breaking`; no blockers)
- Merge #10464 "Enable shellcheck for external test scripts" (just a refactor; no blockers)
- Merge #10429 "Fix and enable external tests" (this PR)
    - Current blocker: flaky Gnosis and OpenZeppelin tests
- Merge #10466 "Running external tests nightly" (cleanup PR)
- Merge #10465 "[BREAKING] Update external tests for 0.8.0" (external tests on `breaking`)
    - Current blocker: failing OpenZeppelin tests.
- Update `_070` and `_080` branches in `solidity-external-tests`
- Merge #10500 "Restore original external test branches" (cleanup PR)

**Comment by cameel** [comment]
I just pushed the final version of this PR and I'm finally marking it as ready to review.

The random failures in OpenZeppelin disappeared after pulling upstream changes into the external test repo. Gnosis is a harder case. I fixed the invalid owner key problem (and submitted a patch upstream: https://github.com/gnosis/safe-contracts/pull/234) but I have no solution for `ProviderError`. What's worse I can't reproduce it locally (even after trying in CircleCI's `node` docker image). I reported it in https://github.com/gnosis/safe-contracts/issues/216. For now I'm just disabling Gnosis v1 test because it's been a big a time-sink and debugging this fully would require even more time.

More details in the updated description. Please take a look at it before reviewing/merging.

File: test/externalTests/gnosis.sh
Comment (hrkrshnn) [inline test/externalTests/gnosis.sh:44]: Why?

File: test/externalTests/gnosis-v2.sh
Comment (hrkrshnn) [inline test/externalTests/gnosis-v2.sh]: These two lines were added as a hacky fix to some problems with node while running tests. Not sure if they are still relevant.

**Review by hrkrshnn — COMMENTED** [review]
Looks good. Two minor comments.

File: test/externalTests/gnosis-v2.sh
Comment (cameel) [reply test/externalTests/gnosis-v2.sh]: Good point. They shouldn't be. These files do not even exist before `run_install` anyway.

I remember that I was confused why it's even being one and I was going to remove it but I could not find it later because I assumed that I saw that part in `common.sh` :) I'll remove it now.

Comment (cameel) [reply test/externalTests/gnosis-v2.sh]: Actually, `package-lock.json` was in Gnosis repo before I pulled in upstream changes and it's generally a good practice to commit it as it pins versions of all the transitive dependencies. Here it probably interfered with the change that forced a different Truffle version. That's no longer needed so we can safely remove it.

Comment (hrkrshnn) [reply test/externalTests/gnosis-v2.sh]: Here is the PR that led to this change: https://github.com/ethereum/solidity/pull/9493

File: test/externalTests/gnosis.sh
Comment (cameel) [reply test/externalTests/gnosis.sh:44]: Because we're no longer forcing it to use Truffle 5.0.42 and the newer one already uses `solc.compile` rather than `solc.compileStandard` so replacing it is not needed.

Maybe I should also remove the `replace_libsolc_call` function from `common.sh` as it's no longer used...

File: test/externalTests/gnosis-v2.sh
Comment (cameel) [reply test/externalTests/gnosis-v2.sh]: Thanks. Yeah, so that's what I assumed. Removing `package.json` was the right thing to do in that situation (it by design overrides `package.json` and is meant o be used only if you want to reproduce the same exact dependency tree the upstream had).

File: test/externalTests/gnosis.sh
Comment (cameel) [reply test/externalTests/gnosis.sh:44]: `replace_libsolc_call` removed.

File: test/externalTests/gnosis-v2.sh
Comment (cameel) [reply test/externalTests/gnosis-v2.sh]: Done. I got rid or removing `node_modules` but I actually left the removal of `package-lock.json` in, at least in those scripts that do modify `package.json`. This makes them a bit more robust because even if the file is not there now, it could always get added to the repo later. I added a comment explaining why it's being done.

**Comment by cameel** [comment]
~This PR is now on top of #10512 because it would conflict otherwise anyway.~ (merged)

**Review by chriseth — APPROVED** [review]

**Pull request merged**
